bfvmm/include/domain/domain.h

```diff
@@ -0,0 +1,214 @@
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef BFVMM_STATIC_DOMAIN
+#ifdef BFVMM_SHARED_DOMAIN
+#define BFVMM_EXPORT_DOMAIN EXPORT_SYM
+#else
+#define BFVMM_EXPORT_DOMAIN IMPORT_SYM
+#endif
+#else
+#define BFVMM_EXPORT_DOMAIN
+#endif
+
+// -----------------------------------------------------------------------------
+// Definitions
+// -----------------------------------------------------------------------------
+
+namespace bfvmm
+{
+
+/// Domain
+///
+class BFVMM_EXPORT_DOMAIN domain : public bfobject
+{
+public:
+
+    /// domain id type
+    ///
+    ///
+    using domainid_type = uint64_t;
+
+public:
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param domainid Id for this domain
+    ///
+    domain(domainid_type domainid);
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    virtual ~domain() = default;
+
+    /// Run
+    ///
+    /// @expects none
+    /// @ensures none
+    ///
+    /// @param obj object that can be passed around as needed
+    ///     by extensions of Bareflank
+    ///
+    VIRTUAL void run(bfobject *obj = nullptr);
+
+    /// Halt
+    ///
+    /// @expects none
+    /// @ensures none
+    ///
+    /// @param obj object that can be passed around as needed
+    ///     by extensions of Bareflank
+    ///
+    VIRTUAL void hlt(bfobject *obj = nullptr);
+
+    /// Init vCPU
+    ///
+    /// @expects none
+    /// @ensures none
+    ///
+    /// @param obj object that can be passed around as needed
+    ///     by extensions of Bareflank
+    ///
+    VIRTUAL void init(bfobject *obj = nullptr);
+
+    /// Fini vCPU
+    ///
+    /// @expects none
+    /// @ensures none
+    ///
+    /// @param obj object that can be passed around as needed
+    ///     by extensions of Bareflank
+    ///
+    VIRTUAL void fini(bfobject *obj = nullptr);
+
+    /// Get ID
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the domain ID
+    ///
+    domainid_type id() const noexcept;
+
+    /// Generate Domain ID
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns a new, unique domain id
+    ///
+    static domainid_type generate_domainid()
+    {
+        static domainid_type s_id = 1;
+        return s_id++;
+    }
+
+    /// Add vCPU
+    ///
+    /// Add a vCPU to this domain
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param vcpu The vcpu to add to this domain
+    void add_vcpu(vcpu_t vcpu)
+    {
+        m_vcpus.push_back(vcpu);
+    }
+
+    /// vcpus
+    ///
+    /// Return the list of vcpus that belong to this domain
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return List of vcpus
+    ///
+    std::list<vcpu_t>
+    vcpus()
+    {
+        return m_vcpus;
+    }
+
+private:
+
+    /// @cond
+
+    domainid_type m_id;
+    std::list<vcpu_t> m_vcpus;
```

Comment:
Why a list and not a deque?

We should consider removing the vcpu_manager and moving it to the domain. Storing the vCPUs in the domain creates a weak reference which means we would need to ensure that a domain ALWAYs out lives a vCPU which this does not. In other words, you could end up with invalid weak references which breaks the ownership model. Technically, a design like this would require the use of shared pointers, not unique pointers, which is not the direction I would like to take. Instead, we should discuss who owns the vCPUs. If the domain owns the vCPUs, the management of the vCPUs should be in the domain. If the domain doesn't own the vCPUs, it should not be storing a list of them. You should ask the owner. Otherwise, the lifetime of the vCPU is hard to reason about which is a C thing, not a modern C++ thing

Reply:
Agreed, not the best implementation, but its a start. I'd rather see a separate PR dedicated to figuring out ownership of vCPUs, because I think we'll run into some complexities there.